File: tests_end_to_end/page_objects/ProjectsPage.py
from playwright.sync_api import Page, expect
import time
from .BasePage import BasePage


class ProjectsPage(BasePage):
    def __init__(self, page: Page):
        super().__init__(page, "projects")
        self.projects_table = self.page.get_by_role("table")

    def click_project(self, project_name):
        self.page.get_by_role("link", name=project_name).click()

    def search_project(self, project_name):
        self.page.get_by_test_id("search-input").click()
        self.page.get_by_test_id("search-input").fill(project_name)

    def check_project_exists_on_current_page(self, project_name):
        expect(
            self.page.get_by_role("cell", name=project_name, exact=True)
        ).to_be_visible(timeout=3)

    def check_project_not_exists_on_current_page(self, project_name):
        expect(
            self.page.get_by_role("cell", name=project_name, exact=True)
        ).not_to_be_visible()

    def check_project_exists_on_current_page_with_retry(self, project_name, timeout):
        start_time = time.time()
        while time.time() - start_time < timeout:
            try:
                self.check_project_exists_on_current_page(project_name=project_name)
                break
            except AssertionError:
                self.page.wait_for_timeout(500)
        else:
            raise AssertionError(
                f"project {project_name} not found in projects list within {timeout} seconds"
            )

    def create_new_project(self, project_name):
        self.page.get_by_role("button", name="Create new project").first.click()

        project_name_fill_box = self.page.get_by_placeholder("Project name")
        project_name_fill_box.click()
        project_name_fill_box.fill(project_name)

        self.page.get_by_role("button", name="Create project").click()

    def delete_project_by_name(self, project_name):
        self.search_project(project_name)
        row = (
            self.page.get_by_role("row")
            .filter(has_text=project_name)
            .filter(has=self.page.get_by_role("cell", name=project_name, exact=True))
        )
        row.get_by_role("button").click()
        self.page.get_by_role("menuitem", name="Delete").click()
        self.page.get_by_role("button", name="Delete project").click()
        self.check_project_not_exists_on_current_page(project_name)

File: tests_end_to_end/tests_end_to_end_ts/typescript-tests/page-objects/projects.page.ts
import { Page, expect } from '@playwright/test';
import { BasePage } from './base.page';

export class ProjectsPage extends BasePage {
  constructor(page: Page) {
    super(page, 'projects');
  }

  async clickProject(projectName: string): Promise<void> {
    await this.page.getByRole('link', { name: projectName }).click();
  }

  async searchProject(projectName: string): Promise<void> {
    await this.page.getByTestId('search-input').click();
    await this.page.getByTestId('search-input').fill(projectName);
  }

  async checkProjectExists(projectName: string): Promise<void> {
    await expect(
      this.page.getByRole('cell', { name: projectName, exact: true })
    ).toBeVisible({ timeout: 3000 });
  }

  async checkProjectNotExists(projectName: string): Promise<void> {
    await expect(
      this.page.getByRole('cell', { name: projectName, exact: true })
    ).not.toBeVisible();
  }

  async checkProjectExistsWithRetry(
    projectName: string,
    timeout: number = 5000
  ): Promise<void> {
    await expect(
      this.page.getByRole('cell', { name: projectName, exact: true })
    ).toBeVisible({ timeout });
  }

  async createNewProject(projectName: string): Promise<void> {
    await this.page.getByRole('button', { name: 'Create new project' }).first().click();

    const projectNameInput = this.page.getByPlaceholder('Project name');
    await projectNameInput.click();
    await projectNameInput.fill(projectName);

    await this.page.getByRole('button', { name: 'Create project' }).click();
  }

  async deleteProjectByName(projectName: string): Promise<void> {
    await this.searchProject(projectName);

    const row = this.page
      .getByRole('row')
      .filter({ hasText: projectName })
      .filter({ has: this.page.getByRole('cell', { name: projectName, exact: true }) });

    await row.getByRole('button').click();
    await this.page.getByRole('menuitem', { name: 'Delete' }).click();
    await this.page.getByRole('button', { name: 'Delete project' }).click();

    await this.checkProjectNotExists(projectName);
  }
}
